src/uu/env/src/split_iterator.rs

```diff
@@ -0,0 +1,505 @@
+
+        self.check_variable_name_start()?;
+
+        loop {
+            match self.get_current_char() {
+                None => break,
+                Some(c) if c.is_ascii_alphanumeric() || c == b'_' => {
+                    self.get_parser_mut().skip_one()?;
+                }
+                Some(_) => break,
+            };
+        }
+
+        let pos_end = self.get_parser().get_look_at_pos();
+
+        if pos_end == pos_start {
+            return Err(ParseError::ParsingOfVariableNameFailed {
+                pos: pos_start,
+                msg: "Missing variable name".into(),
+            });
+        }
+
+        Ok(self.get_parser().get_substring(&Range {
+            start: pos_start,
+            end: pos_end,
+        })?)
+    }
+
+    fn substitute_variable(&mut self) -> Result<(), ParseError> {
+        self.get_parser_mut().skip_one()?;
+
+        let (name, default) = match self.get_current_char() {
+            None => {
+                return Err(ParseError::ParsingOfVariableNameFailed {
+                    pos: self.get_parser().get_look_at_pos(),
+                    msg: "missing variable name".into(),
+                })
+            }
+            Some(b'{') => {
+                self.skip_one()?;
+                self.parse_braced_variable_name()?
+            }
+            Some(_) => (self.parse_unbraced_variable_name()?, None),
+        };
+
+        let value = std::env::var(name).ok();
+        match (&value, default) {
+            (None, None) => {} // do nothing, just replace it with ""
+            (Some(value), _) => {
+                self.raw_parser.put_string_utf8(value)?;
+            }
+            (None, Some(default)) => {
+                self.raw_parser.put_string_utf8(default)?;
+            }
+        };
+
+        Ok(())
+    }
+
+    fn check_and_replace_ascii_escape_code(&mut self, c: u8) -> Result<bool, ParseError> {
+        let (from, to) = REPLACEMENTS;
+        if let Some(pos) = memchr::memchr(c, from) {
+            self.skip_one()?;
+            self.push_ascii_char_to_word(*to.get(pos).unwrap())?;
+            return Ok(true);
+        }
+
+        Ok(false)
+    }
+
+    fn make_invalid_sequence_backslash_xin_minus_s(&self, c: u8) -> ParseError {
+        let valid_char: char = if is_ascii(c) {
+            c.into()
+        } else {
+            INVALID_UTF8_MARKER
+        };
+        ParseError::InvalidSequenceBackslashXInMinusS {
+            pos: self.raw_parser.get_parser().get_look_at_pos(),
+            c: valid_char,
+        }
+    }
+
+    pub fn split(&mut self) -> Result<Vec<String>, ParseError> {
+        use State::*;
+
+        loop {
+            let c = self.get_current_char();
+            let _c_char =
+                c.map(|c| -> char { char::from_u32(c.into()).unwrap_or(INVALID_UTF8_MARKER) }); // just for debugging session. In release, compiler will remove
+            self.state = match self.state {
+                Delimiter => match c {
```

Comment:
I'm not really a fan of this style of parsing. I find that it's often easier to implement parsers in a way where the control flow is the state. I'll try to explain what I mean. Say we parse single quotes. The structure would be:
```rust
loop {
    let c = current_char();
    match c {
        '\' => {
            // parse everything until the closing quote
            // in this block
        }
    }
}
```
Instead of moving to a `SingleQuoted` state. I find that easier to reason about, because the start of the quotes and the rest of it's parsing logic are closer together, but maybe that's different for you. If you have some arguments for this style we can keep it.

Reply:
done